docs/gcloud-auth.rst

```diff
@@ -0,0 +1,264 @@
+.. toctree::
+  :maxdepth: 1
+  :hidden:
+
+Authentication
+--------------
+
+==============
+Quick overview
+==============
+
+**If you're running in Compute Engine or App Engine**,
+authentication should "just work"
+as credentials can be loaded automatically
+and therefore there is no extra code required to authenticate.
+
+**If you're developing locally**,
+the easiest way to authenticate is using the Cloud SDK.
+
+**If you're running your application elsewhere**,
+you should download the service account JSON keyfile
+and point to it using an environment variable.
+
+-----------------------------------------------
+In what order does gcloud discover credentials?
+-----------------------------------------------
+
+``gcloud`` will look in a bunch of different places
+for authentication credentials, in the following order:
+
+#. Credentials you explicitly set in code
+#. Credentials from an environment variable (``GOOGLE_APPLICATION_CREDENTIALS``)
+#. Credentials provided by the Cloud SDK (``gcloud auth login``)
+#. Credentials provided when running in Google App Engine
+#. Credentials provided when running in Google Compute Engine
+
+This means that you can set your credentials a bunch of different ways.
+We recommend using some form of implicitly discovered credentials
```

Comment:
The credentials are thread safe. Only httplib2.Http object is the problem. It's easily avoidable with using thread local Http object.
However, if you use multiprocessing, this approach sometimes breaks. (Maybe there is a workaround)

Anyways, is it possible for gcloud to always use thread local Http object? Then multithreaded code authors can just use the implicit method.